docs/features/source-generators.md

```diff
@@ -28,102 +28,114 @@ namespace Microsoft.CodeAnalysis
 {
     public interface ISourceGenerator
     {
+        void Initialize(InitializationContext context);
         void Execute(SourceGeneratorContext context);
     }
 }
 ```
 
 Generator implementations are defined in external assemblies passed to the compiler
-using the same `-analyzer:` option used for diagnostic analyzers.
+using the same `-analyzer:` option used for diagnostic analyzers. Implementations are required to
+be annotated with a `Microsoft.CodeAnalysis.GeneratorAttribute` attribute
+in order to aid discovery.
 
 An assembly can contain a mix of diagnostic analyzers and source generators.
 Since generators are loaded from external assemblies, a generator cannot be used to build
 the assembly in which it is defined.
 
-`ISourceGenerator` has a single `Execute` method that is called by the host (either the IDE
-or the command-line compiler). `Execute` passes an instance of `SourceGeneratorContext` that provides access 
-to the `Compilation` and allows the generator to alter it by adding source and reporting diagnostics.
+`ISourceGenerator` has an `Initialize` method that is called by the host (either the IDE or
+the command-line compiler) exactly once.`Initialize` passes an instance of `InitializationContext`
+which can be used by the generator to register a set of callbacks that affect how future generation
+passes will occur.
+
+The main generation pass occurs via the `Execute` method. `Execute` passes an instance of `SourceGeneratorContext`
+that provides access to the `Compilation` and allows the generator to alter it by adding source
```

Comment:
>it [](start = 76, length = 2)

"the `SourceGeneratorContext`"? Consider being explicit because it's not clear if "it" refers to the `SourceGeneratorContext` or the `Compilation`.

Reply:
Perhaps a minor point, but the generator is not directly altering the `Compilation` because it looks like the generator sees a single fixed `Compilation` before any generators run. On the other hand, I'm not sure how to express this differently.

---
In reply to: [416903004](https://github.com/dotnet/roslyn/pull/43753#discussion_r416903004) [](ancestors = 416903004)

Reply:
This is good feedback. I've tried to clarify it a bit by referring to the current and output compilations (also added additional files, because somehow that was apparently missed)

---
In reply to: [416924714](https://github.com/dotnet/roslyn/pull/43753#discussion_r416924714) [](ancestors = 416924714,416903004)